File: lib-util/src/main/java/com/ustadmobile/core/impl/UmCallbackWithDefaultValue.java
package com.ustadmobile.core.impl;

public class UmCallbackWithDefaultValue<T> implements UmCallback<T>{

    private T defaultVal;

    private UmCallback<T> callback;

    public UmCallbackWithDefaultValue(T defaultVal, UmCallback callback) {
        this.defaultVal = defaultVal;
        this.callback = callback;
    }

    @Override
    public void onSuccess(T result) {
        if(result != null)
            UmCallbackUtil.onSuccessIfNotNull(callback, result);
        else
            UmCallbackUtil.onSuccessIfNotNull(callback, defaultVal);
    }

    @Override
    public void onFailure(Throwable exception) {
        UmCallbackUtil.onFailIfNotNull(callback, exception);
    }
}

File: lib-util/src/main/java/com/ustadmobile/core/impl/UmCallbackWithDefaultValue.kt
package com.ustadmobile.core.impl

class UmCallbackWithDefaultValue<T>(private val defaultVal: T, callback: UmCallback<*>) : UmCallback<T> {

    private val callback: UmCallback<T> = callback as UmCallback<T>

    override fun onSuccess(result: T) {
        if (result != null)
            UmCallbackUtil.onSuccessIfNotNull(callback, result)
        else
            UmCallbackUtil.onSuccessIfNotNull(callback, defaultVal)
    }

    override fun onFailure(exception: Throwable) {
        UmCallbackUtil.onFailIfNotNull(callback, exception)
    }
}
